src/lib/storage/table.cpp

```diff
@@ -17,53 +19,103 @@
+  DebugAssert(search == _column_name_mapping.end(), "Column names may not be duplicated");
+  _column_name_mapping.insert({name, (ColumnID)_column_names.size()});
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+  auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(type);
+  _chunks.front()->add_segment(segment);
 }
 
 void Table::append(std::vector<AllTypeVariant> values) {
-  // Implementation goes here
-}
+  // Check if the latest Chunk has already reached maximum size and create a new one if necessary
+  if (this->is_new_chunk_needed()) {
+    _chunks.push_back(std::make_shared<Chunk>());
 
-uint16_t Table::column_count() const {
-  // Implementation goes here
-  return 0;
+    for (auto type_name : _column_types) {
+      auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(type_name);
+      _chunks.back()->add_segment(segment);
+    }
+  }
+
+  _chunks.back()->append(values);
 }
 
+bool Table::is_new_chunk_needed() { return _chunks.back()->size() >= _max_chunk_size; }
+
+uint16_t Table::column_count() const { return _column_names.size(); }
+
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  // Since we cannot delete data, all old chunks will have maximum chunk size
+  return ((_chunks.size() - 1) * _max_chunk_size) + _chunks.back()->size();
 }
 
-ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return (ChunkID)_chunks.size(); }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
+  auto search = _column_name_mapping.find(column_name);
+  DebugAssert(search != _column_name_mapping.end(), "Could not find column_name");
+  return search->second;
 }
 
-uint32_t Table::max_chunk_size() const {
-  // Implementation goes here
-  return 0;
-}
+uint32_t Table::max_chunk_size() const { return _max_chunk_size; }
 
-const std::vector<std::string>& Table::column_names() const {
-  throw std::runtime_error("Implement Table::column_names()");
-}
+bool Table::is_empty() { return _chunks.size() == 1 && _chunks.front()->size() == 0; }
 
-const std::string& Table::column_name(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_name");
-}
+const std::vector<std::string>& Table::column_names() const { return _column_names; }
 
-const std::string& Table::column_type(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_type");
-}
+const std::string& Table::column_name(ColumnID column_id) const { return _column_names.at(column_id); }
+
+const std::string& Table::column_type(ColumnID column_id) const { return _column_types.at(column_id); }
+
+Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks.at(chunk_id); }
+
+const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
+
+void Table::compress_chunk(ChunkID chunk_id) {
```

Comment:
This function is called in tests, but it is not actually tested if it succeeds, right?

Reply:
according to the assignment, thread safety mechanisms should be employed in for the chunk compression